core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -360,6 +371,17 @@ class BlockManagerMasterEndpoint(
 
   }
 
+  private def addMergerLocation(blockManagerId: BlockManagerId): Unit = {
+    if (!shuffleMergerLocations.contains(blockManagerId.host) && !blockManagerId.isDriver) {
+      val shuffleServerId = BlockManagerId(blockManagerId.executorId, blockManagerId.host,
+        StorageUtils.externalShuffleServicePort(conf))
+      if (shuffleMergerLocations.size >= maxRetainedMergerLocations) {
+        shuffleMergerLocations -= shuffleMergerLocations.head._1
+      }
+      shuffleMergerLocations(shuffleServerId.host) = shuffleServerId
```

Comment:
I wonder why you need a separate `shuffleMergerLocations` since it doesn't have a big difference comparing to the `BlockManagerId`. I guess this is because of the dynamic allocation case where a node may have no running executors but external shuffle service only...right?

Reply:
Yes, thats right. this is to keep track of locations where an executor ran before that we we know the application has registered with that particular host's External shuffle service which is needed to push the shuffle data.